File: Examples/GroupDocs.Metadata.Examples.VBasic/APIs.vb
Imports GroupDocs.Metadata.Examples.VBasic.Utilities 
Imports GroupDocs.Metadata
Imports GroupDocs.Metadata.Formats.Image
Imports GroupDocs.Metadata.Tools
Imports System.Collections.Generic
Imports System.Data
Imports System.IO
Imports System.Linq
Imports System.Text

Namespace GroupDocs.Metadata.Examples.VBasic
    Public NotInheritable Class APIs
        Private Sub New()
        End Sub
        Public NotInheritable Class Document
            Private Sub New()
            End Sub
            ''' <summary>
            ''' Compares metadata of two documents and displays result 
            ''' </summary> 
            Public Shared Sub CompareDocument(firstDocument As String, secondDocument As String, type As ComparerSearchType)
                Try
                    'ExStart:ComparisonAPI
                    firstDocument = Common.MapSourceFilePath(firstDocument)
                    secondDocument = Common.MapSourceFilePath(secondDocument)

                    Dim differences As MetadataPropertyCollection = ComparisonFacade.CompareDocuments(firstDocument, secondDocument, type)

                    For Each [property] As MetadataProperty In differences
                        Console.WriteLine("{0} : {1}", [property].Name, [property].Value)
                        'ExEnd:ComparisonAPI
                    Next
                Catch exp As Exception
                    Console.WriteLine("Exception occurred: " + exp.Message)
                End Try

            End Sub
            ''' <summary>
            ''' Searches metadata in document 
            ''' </summary> 
            Public Shared Sub SearchMetadata(filePath As String, propertyName As String, searchCondition As SearchCondition)
                Try
                    'ExStart:DocumentSearchAPI
                    filePath = Common.MapSourceFilePath(filePath)

                    Dim properties As MetadataPropertyCollection = SearchFacade.ScanDocument(filePath, propertyName, searchCondition)

                    For Each [property] As MetadataProperty In properties
                        Console.WriteLine("{0} : {1}", [property].Name, [property].Value)
                        'ExEnd:DocumentSearchAPI
                    Next
                Catch exp As Exception
                    Console.WriteLine("Exception occurred: " + exp.Message)
                End Try

            End Sub

            ''' <summary>
            ''' Replaces author name in document using custom ReplaceHandler
            ''' </summary> 
            Public Shared Sub ReplaceAuthorName(filePath As String)
                Try
                    'ExStart:ReplaceAuthorName
                    ' initialize custom handler, send output path using constructor
                    Dim replaceHandler As IReplaceHandler(Of MetadataProperty) = New AuthorReplaceHandler(Common.MapDestinationFilePath(filePath))

                    ' replace author
                    'ExEnd:ReplaceAuthorName
                    Dim affectedPropertiesCount As Integer = SearchFacade.ReplaceInDocument(Common.MapSourceFilePath(filePath), replaceHandler)
                Catch exp As Exception
                    Console.WriteLine("Exception occurred: " + exp.Message)
                End Try

            End Sub
            ''' <summary>
            ''' Replaces author name in document using custom ReplaceHandler
            ''' </summary> 
            Public Shared Sub ReplaceMetadataProperties(filePath As String)
                Try
                    'ExStart:ReplaceMetadataProperties
                    ' replace 'author' value
                    SearchFacade.ReplaceInDocument(Common.MapSourceFilePath(filePath), "Author", "Jack London", SearchCondition.Matches, Common.MapDestinationFilePath(filePath))

                    ' replace all properties contained 'co' to 'some value'
                    SearchFacade.ReplaceInDocument(Common.MapSourceFilePath(filePath), "co", "some value", SearchCondition.Contains, Common.MapDestinationFilePath(filePath))
                    'ExEnd:ReplaceMetadataProperties
                Catch exp As Exception
                    Console.WriteLine("Exception occurred: " + exp.Message)
                End Try

            End Sub
        End Class
        Public NotInheritable Class Image
            Private Sub New()
            End Sub
            ''' <summary>
            ''' Searches metadata in image 
            ''' </summary> 
            Public Shared Sub SearchMetadata(filePath As String, propertyName As String, searchCondition As SearchCondition)
                Try
                    'ExStart:ImageSearchAPI
                    filePath = Common.MapSourceFilePath(filePath)

                    ' looking the software
                    Dim properties As ExifProperty() = SearchFacade.ScanExif(filePath, propertyName, searchCondition)

                    For Each [property] As ExifProperty In properties
                        Console.WriteLine("{0} : {1}", [property].Name, [property].ToString())
                    Next
                    'ExEnd:ImageSearchAPI
                Catch exp As Exception
                    Console.WriteLine("Exception occurred: " + exp.Message)
                End Try

            End Sub

            ''' <summary>
            ''' Compares EXIF metadata of two jpeg files 
            ''' </summary> 
            Public Shared Sub CompareExifMetadata(firstFile As String, secondFile As String, type As ComparerSearchType)
                Try
                    'ExStart:ExifComparisonAPI
                    firstFile = Common.MapSourceFilePath(firstFile)
                    secondFile = Common.MapSourceFilePath(secondFile)

                    Dim differences As ExifProperty() = ComparisonFacade.CompareExif(firstFile, secondFile, type)

                    For Each [property] As ExifProperty In differences
                        Console.WriteLine("{0} : {1}", [property].Name, [property].ToString())
                    Next
                    'ExEnd:ExifComparisonAPI
                Catch exp As Exception
                    Console.WriteLine("Exception occurred: " + exp.Message)
                End Try

            End Sub
        End Class

        ''' <summary>
        ''' Exports metadata of specified file into specified type 
        ''' </summary> 
        Public Shared Sub ExportMetadata(filePath As String, exportType As Integer)
            Try
                'ExStart:ExportMetadataAPI
                filePath = Common.MapSourceFilePath(filePath)

                If exportType = ExportTypes.ToExcel Then
                    'ExStart:ExportMetadataToExcel
                    ' path to the output file
                    Dim outputPath As String = Common.MapDestinationFilePath("metadata.xlsx")

                    ' export to excel
                    Dim content As Byte() = ExportFacade.ExportToExcel(filePath)

                    ' write data to the file
                    File.WriteAllBytes(outputPath, content)
                    'ExEnd:ExportMetadataToExcel
                ElseIf exportType = ExportTypes.ToCSV Then
                    'ExStart:ExportMetadataToCSV
                    ' path to the output file
                    Dim outputPath As String = Common.MapDestinationFilePath("metadata.csv")

                    ' export to csv
                    Dim content As Byte() = ExportFacade.ExportToCsv(filePath)

                    ' write data to the file
                    File.WriteAllBytes(outputPath, content)
                    'ExEnd:ExportMetadataToCSV
                Else
                    'ExStart:ExportMetadataToDataSet
                    ' export to DataSet
                    Dim ds As DataSet = ExportFacade.ExportToDataSet(filePath)
                    ' get first table
                    Dim products As DataTable = ds.Tables(0)

                    ' need to System.Data.DataSetExtension reference            
                    Dim query As IEnumerable(Of DataRow) = From product In products.AsEnumerable()

                    Console.WriteLine("Properties:")
                    For Each p As DataRow In query
                        Console.Write(p.Field(Of String)("Metadata property"))
                        Console.Write(": ")
                        Console.WriteLine(p.Field(Of String)("Value"))
                    Next
                    'ExEnd:ExportMetadataToDataSet
                    'ExEnd:ExportMetadataAPI
                End If
            Catch exp As Exception
                Console.WriteLine("Exception occurred: " + exp.Message)
            End Try

        End Sub
    End Class
End Namespace

File: Examples/GroupDocs.Metadata.Examples.CSharp/APIs.cs
﻿using GroupDocs.Metadata.Examples.CSharp.Utilities;
using GroupDocs.Metadata.Examples.Utilities.CSharp;
using GroupDocs.Metadata;
using GroupDocs.Metadata.Formats.Image;
using GroupDocs.Metadata.Tools;
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using GroupDocs.Metadata.Formats;

namespace GroupDocs.Metadata.Examples.CSharp
{
    public static class APIs
    {
        public static class Document
        {
            /// <summary>
            /// Compares metadata of two documents and displays result 
            /// </summary> 
            public static void CompareDocument(string firstDocument, string secondDocument, ComparerSearchType type)
            {
                try
                {
                    //ExStart:ComparisonAPI
                    firstDocument = Common.MapSourceFilePath(firstDocument);
                    secondDocument = Common.MapSourceFilePath(secondDocument);

                    MetadataPropertyCollection differences = ComparisonFacade.CompareDocuments(firstDocument, secondDocument, type);

                    foreach (MetadataProperty property in differences)
                    {
                        Console.WriteLine("{0} : {1}", property.Name, property.Value);
                    }
                    //ExEnd:ComparisonAPI
                }
                catch (Exception exp)
                {
                    Console.WriteLine("Exception occurred: " + exp.Message);
                }

            }
            /// <summary>
            /// Searches metadata in document 
            /// </summary> 
            public static void SearchMetadata(string filePath, string propertyName, SearchCondition searchCondition)
            {
                try
                {
                    //ExStart:DocumentSearchAPI
                    filePath = Common.MapSourceFilePath(filePath);

                    MetadataPropertyCollection properties = SearchFacade.ScanDocument(filePath, propertyName, searchCondition);

                    foreach (MetadataProperty property in properties)
                    {
                        Console.WriteLine("{0} : {1}", property.Name, property.Value);
                    }
                    //ExEnd:DocumentSearchAPI
                }
                catch (Exception exp)
                {
                    Console.WriteLine("Exception occurred: " + exp.Message);
                }

            }

            /// <summary>
            /// Replaces author name in document using custom ReplaceHandler
            /// </summary> 
            public static void ReplaceAuthorName(string filePath)
            {
                try
                {
                    //ExStart:ReplaceAuthorName
                    // initialize custom handler, send output path using constructor
                    IReplaceHandler<MetadataProperty> replaceHandler = new AuthorReplaceHandler(Common.MapDestinationFilePath(filePath));

                    // replace author
                    int affectedPropertiesCount = SearchFacade.ReplaceInDocument(Common.MapSourceFilePath(filePath), replaceHandler);
                    //ExEnd:ReplaceAuthorName
                }
                catch (Exception exp)
                {
                    Console.WriteLine("Exception occurred: " + exp.Message);
                }

            }
            /// <summary>
            /// Replaces author name in document using custom ReplaceHandler
            /// </summary> 
            public static void ReplaceMetadataProperties(string filePath)
            {
                try
                {
                    //ExStart:ReplaceMetadataProperties
                    // replace 'author' value
                    SearchFacade.ReplaceInDocument(Common.MapSourceFilePath(filePath), "Author", "Jack London", SearchCondition.Matches, Common.MapDestinationFilePath(filePath));

                    // replace all properties contained 'co' to 'some value'
                    SearchFacade.ReplaceInDocument(Common.MapSourceFilePath(filePath), "co", "some value", SearchCondition.Contains, Common.MapDestinationFilePath(filePath));
                    //ExEnd:ReplaceMetadataProperties
                }
                catch (Exception exp)
                {
                    Console.WriteLine("Exception occurred: " + exp.Message);
                }

            }
        }
        public static class Image
        {
            /// <summary>
            /// Searches metadata in image 
            /// </summary> 
            public static void SearchMetadata(string filePath, string propertyName, SearchCondition searchCondition)
            {
                try
                {
                    //ExStart:ImageSearchAPI
                    filePath = Common.MapSourceFilePath(filePath);

					TiffTag[] resolutionTags = SearchFacade.ScanExifTags(filePath, new Regex("^(XResolution|YResolution)$"));

					foreach (TiffTag resolutionTag in resolutionTags)
					{
						Console.WriteLine("{0} - {1}", resolutionTag.DefinedTag, ((TiffRationalTag)resolutionTag).TagValue[0].Value);
					}
					//ExEnd:ImageSearchAPI
				}
                catch (Exception exp)
                {
                    Console.WriteLine("Exception occurred: " + exp.Message);
                }

            }

            /// <summary>
            /// Compares EXIF metadata of two jpeg files 
            /// </summary> 
            public static void CompareExifMetadata(string firstFile, string secondFile, ComparerSearchType type)
            {
                try
                {
                    //ExStart:ExifComparisonAPI
                    firstFile = Common.MapSourceFilePath(firstFile);
                    secondFile = Common.MapSourceFilePath(secondFile);

					TiffTag[] intersection = ComparisonFacade.CompareExifTags(firstFile, secondFile, ComparerSearchType.Intersection);

					foreach (TiffTag tag in intersection)
					{
						Console.WriteLine(tag.DefinedTag);
						Console.WriteLine(tag.TagType);
						Console.WriteLine(tag.GetFormattedValue());
					}
					//ExEnd:ExifComparisonAPI
				}
                catch (Exception exp)
                {
                    Console.WriteLine("Exception occurred: " + exp.Message);
                }

            }
        }

        /// <summary>
        /// Exports metadata of specified file into specified type 
        /// </summary> 
        public static void ExportMetadata(string filePath, int exportType)
        {
            try
            {
                //ExStart:ExportMetadataAPI
                filePath = Common.MapSourceFilePath(filePath);

                if (exportType == ExportTypes.ToExcel)
                {
                    //ExStart:ExportMetadataToExcel
                    // path to the output file
                    string outputPath = Common.MapDestinationFilePath("metadata.xlsx"); 

                    // export to excel
                    byte[] content = ExportFacade.ExportToExcel(filePath);

                    // write data to the file
                    File.WriteAllBytes(outputPath, content);
                    //ExEnd:ExportMetadataToExcel
                }
                else if (exportType == ExportTypes.ToCSV)
                {
                    //ExStart:ExportMetadataToCVS
                    // path to the output file
                    string outputPath = Common.MapDestinationFilePath("metadata.csv"); 

                    // export to csv
                    byte[] content = ExportFacade.ExportToCsv(filePath);

                    // write data to the file
                    File.WriteAllBytes(outputPath, content);
                    //ExEnd:ExportMetadataToCVS
                }
                else  
                {
                    //ExStart:ExportMetadataToDataSet
                    // export to DataSet
                    DataSet ds = ExportFacade.ExportToDataSet(filePath);
                    // get first table
                    DataTable products = ds.Tables[0];

                    // need to System.Data.DataSetExtension reference            
                    IEnumerable<DataRow> query =
                    from product in products.AsEnumerable()
                    select product;

                    Console.WriteLine("Properties:");
                    foreach (DataRow p in query)
                    {
                        Console.Write(p.Field<string>("Metadata property"));
                        Console.Write(": ");
                        Console.WriteLine(p.Field<string>("Value"));
                    }
                    //ExEnd:ExportMetadataToDataSet
                }
                //ExEnd:ExportMetadataAPI
            }
            catch (Exception exp)
            {
                Console.WriteLine("Exception occurred: " + exp.Message);
            }

        }
        /// <summary>
        /// Demonstrate the use of Refactor base metadata classes to add support for hierarchical metadata structures
        /// Feature is supported by version 19.2 or greater
        /// </summary>
        public static void DemonstrateMetadataTree()
        {
            //Source file path
            string filePath = "Video/ASF/sample.asf";

            using (FormatBase format = FormatFactory.RecognizeFormat(Common.MapSourceFilePath(filePath)))
            {

                DisplayMetadataTree(format.GetMetadata(), 0);
            }

        }
        private static void DisplayMetadataTree(Metadata metadata, int indent)
        {
            if (metadata != null)
            {
                var stringMetadataType = metadata.MetadataType.ToString();

                Console.WriteLine(stringMetadataType.PadLeft(stringMetadataType.Length + indent));

                foreach (MetadataProperty property in metadata)
                {
                    string stringPropertyRepresentation = string.Format("Name: {0}, Value: {1}", property.Name, property.GetFormattedValue());

                    Console.WriteLine(stringPropertyRepresentation.PadLeft(stringPropertyRepresentation.Length + indent + 1));

                    if (property.Value != null)
                    {
                        switch (property.Value.Type)
                        {
                            case MetadataPropertyType.Metadata:
                                DisplayMetadataTree(property.Value.ToMetadata(), indent + 2);
                                break;
                            case MetadataPropertyType.MetadataArray:
                                DisplayMetadataTree(property.Value.ToMetadataArray(), indent + 2);
                                break;
                        }
                    }
                }
            }
        }
        private static void DisplayMetadataTree(Metadata[] metadataArray, int indent)
        {

            if (metadataArray != null)
            {
                foreach (Metadata metadata in metadataArray)
                {
                    DisplayMetadataTree(metadata, indent);
                }
            }
        }
    }
}
